tree/dataframe/src/RDFSnapshotHelpers.cxx

```diff
@@ -930,3 +995,244 @@ ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::MakeNew(void *newName)
       fNSlots,           finalName, fDirName,          fNTupleName,        fInputFieldNames,
       fOutputFieldNames, fOptions,  fInputLoopManager, fOutputLoopManager, fInputColumnTypeIDs};
 }
+
+/*
+ * ------------------------------------
+ * Snapshot with systematic variations
+ * ------------------------------------
+ */
+namespace ROOT::Internal::RDF {
+/// An object to store an output file and a tree in one common place to share them between instances
+/// of Snapshot with systematic uncertainties.
+struct SnapshotOutputWriter {
+   std::unique_ptr<TFile> fFile;
+   std::unique_ptr<TTree> fTree;
+   std::string fDirectoryName;
+   RLoopManager *fOutputLoopManager;
+
+   // Bitmasks to indicate whether syst. uncertainties have been computed. Bound to TBranches, so need to be stable in
+   // memory.
+   struct Bitmask {
+      std::string branchName;
+      std::bitset<64> bitset{};
+      std::unique_ptr<uint64_t> branchBuffer{new uint64_t{}};
+   };
+   std::vector<Bitmask> fBitMasks;
+
+   std::unordered_map<std::string, unsigned int> fBranchToVariationMapping;
+   // The corresponding ROOT dictionary is declared in core/clingutils/src
+   std::unordered_map<std::string, std::pair<std::string, unsigned int>> fBranchToBitmaskMapping;
+   unsigned int fNBits = 0;
+
+   SnapshotOutputWriter(TFile *file, TTree *tree) : fFile{file}, fTree{tree} {}
+   ~SnapshotOutputWriter()
+   {
+      if (!fBranchToBitmaskMapping.empty()) {
+         fFile->WriteObject(&fBranchToBitmaskMapping,
+                            (std::string{"R_rdf_branchToBitmaskMapping_"} + fTree->GetName()).c_str());
+      }
```

Comment:
Sorry I'm getting lost here, isn't the mapping another branch of the output tree? Here instead it is an object in the output file

Reply:
It's indeed both. For every 64 variations, we add one branch to the output tree. But to know which variation is which bit in which branch, we write a map that maps branchName --> (branchWithBitmask, index) pairs.
The masked column readers will check for this map, and start the masking logic if their column shows up in this map.